dune/grid/common/p2pcommunicator.hh

```diff
@@ -0,0 +1,240 @@
+  {
+    typedef std::vector< char >  BufferType;
+
+    mutable BufferType buffer_;
+    const double factor_;
+    mutable size_t pos_;
+public:
+    /** \brief constructor taking memory reserve estimation factor (default is 1.1, i.e. 10% over estimation )
+     */
+    SimpleMessageBuffer( const double factor = 1.1 )
+      : buffer_(), factor_( factor )
+    {
+      resetReadPosition();
+    }
+
+    /** \brief clear the buffer */
+    void clear() { BufferType().swap( buffer_ ); resetReadPosition(); }
+    /** \brief reset read position of buffer to beginning */
+    void resetReadPosition() { pos_ = 0 ; }
+    /** \brief return size of buffer */
+    size_t size() const { return buffer_.size(); }
+
+    /** \brief reserve memory for 'size' entries  */
+    void reserve( const size_t size )
+    {
+        buffer_.reserve( size );
+    }
+
+    /** \brief resize buffer to 'size' entries  */
+    void resize( const size_t size )
+    {
+        buffer_.resize( size );
+    }
+
+    /** \brief write value to buffer, value must implement the operator= correctly (i.e. no internal pointers etc.) */
+    template <class T>
+    void write( const T& value )
+    {
+      // union to access bytes in value
+      const size_t tsize = sizeof( T );
+      union { T value; char bytes[ tsize ]; } convert;
+      convert.value = value;
+      size_t pos  = buffer_.size();
+      const size_t sizeNeeded = pos + tsize ;
+      // reserve with some 10% overestimation
+      if( buffer_.capacity() < sizeNeeded )
+      {
+        reserve( size_t(factor_ * sizeNeeded) ) ;
+      }
+      // resize to size need to store value
+      buffer_.resize( sizeNeeded );
+      for(unsigned int i=0; i<tsize; ++i )
+      {
+        buffer_[ pos++ ] = convert.bytes[ i ] ;
+      }
+    }
+
+    /** \brief read value from buffer, value must implement the operator= correctly (i.e. no internal pointers etc.) */
+    template <class T>
+    void read( T& value ) const
+    {
+      // read bytes from stream and store in value
+      const size_t tsize = sizeof( T );
+      union { T value; char bytes[ tsize ]; } convert;
+      assert( pos_ + tsize <= buffer_.size() );
+      for( unsigned int i=0; i<tsize; ++i )
+      {
+        convert.bytes[ i ] = buffer_[ pos_++ ];
+      }
+      value = convert.value;
+    }
+
+    /** \brief return pointer to buffer and size for use with MPI functions */
+    std::pair< char* , int > buffer() const
+    {
+      return std::make_pair( buffer_.data(), int(buffer_.size()) );
+    }
+  };
+
+  /** \brief Point-2-Point communicator for exchange messages between processes */
+  template <class MsgBuffer, class Comm = MPIHelper::MPICommunicator >
+  class Point2PointCommunicator : public CollectiveCommunication< Comm >
+  {
+  protected:
+    typedef CollectiveCommunication< Comm > BaseType;
+    typedef Point2PointCommunicator< MsgBuffer, Comm > ThisType;
+
+    // starting message tag
+    static const int messagetag = 234;
```

Comment:
It's and arbitrary number away from standard MPI message tasks which should be between 0 and 32768.